pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -316,6 +316,7 @@ func (m *managerImpl) processShutdownEvent() error {
 				klog.V(1).InfoS("Shutdown manager killing pod with gracePeriod", "pod", klog.KObj(pod), "gracePeriod", gracePeriodOverride)
 
 				if err := m.killPodFunc(pod, false, &gracePeriodOverride, func(status *v1.PodStatus) {
```

Comment:
> I'm referring to specifically a pod that answers graceful shutdown with exit code 0 that is restartonfailure. That should exit as succeeded (hardcoding Failed in the phase is incorrect).
https://github.com/kubernetes/kubernetes/pull/104798#issuecomment-924066206
/cc @smarterclayton


I had a similar fix before, and I don't have a strong opinion

Reply:
/cc @smarterclayton

Reply:
This should be an if check - `if Phase != <terminal phase>, set phase failed`

Reply:
> This should be an if check - `if Phase != <terminal phase>, set phase failed`

hmm, I don't completely follow. We are setting the pod phase to failed here, so I'm a little confused which phase you mentioned we should be reading here (i.e. the `Phase != <terminal phase>`)

Reply:
@bobbypage I'm not seeing how it's possible to encounter a succeeded pod here, but just as a protection, you can do this:

```go
if status.Phase = v1.PodSucceeded {
	status.Phase = v1.PodFailed
}
```

Reply:
No, you can absolutely get already succeeded pods here.  Two kill requests can happen at the same time, and the pod can already be terminating or successfully finish between when you call killPod and when your callback gets invoked (kill pod is async).  Jordan's suggested code is fine.

Reply:
Thanks for the clarification and suggestion. Updated as Jordan mentioned.

Reply:
This lgtm now in terms of fixing the 1.22 regression while not introducing a Succeeded → Failed transition.

Follow-up in 1.24 is definitely needed for the other scenarios described in https://github.com/kubernetes/kubernetes/pull/104798#issuecomment-924066206 (pod terminates successfully without wanting to be restarted, etc).